pkg/utils/termination/termination.go

```
@@ -0,0 +1,65 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package termination
+
+import (
+	"context"
+	"fmt"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+)
+
+// EnsureTerminated is a helper function that takes a v1beta1.NodeClaim and calls cloudProvider.Delete() if status condition
+// on nodeClaim is not terminating. If it is terminating then it will call cloudProvider.Get() to check if the instance
+// is terminated or not. It will return an error and a boolean that indicates if the instance is terminated or not.
+func EnsureTerminated(ctx context.Context, c client.Client, nodeClaim *v1beta1.NodeClaim, cloudProvider cloudprovider.CloudProvider) (bool, error) {
+	// Check if the status condition on nodeClaim is Terminating
+	if !nodeClaim.StatusConditions().Get(v1beta1.ConditionTypeTerminating).IsTrue() {
+		// If not then call Delete on cloudProvider to trigger termination and always requeue reconciliation
+		if err := cloudProvider.Delete(ctx, nodeClaim); err != nil {
+			if cloudprovider.IsNodeClaimNotFoundError(err) {
+				nodeClaim.StatusConditions().SetTrue(v1beta1.ConditionTypeTerminating)
+				// We call Update() here rather than Patch() because patching a list with a JSON merge patch
+				// can cause races due to the fact that it fully replaces the list on a change
+				// https://github.com/kubernetes/kubernetes/issues/111643#issuecomment-2016489732
+				if err = c.Status().Update(ctx, nodeClaim); err != nil {
```

Should we just return the error as-is without ignoring the NotFound error? That way the caller can handle this NotFound error and bail early?